feat: Year Calendar UI 구현

pro-popo opened the pull request:
✅ `moment.js` 라이브러리 사용

resolves: GC-13 #done

pro-popo committed:
feat: Year Calendar UI 구현
resolves: GC-13 #done

pro-popo commented:
✅ 컴포넌트 설명
- `YearContainer.tsx`
: 해당 년(YEAR)의 월별 캘린더 출력 (1월~12월)
- `MonthContainer.tsx`
: 해당 월(MONTH)의 6개의 주(WEEK) 출력
- `Week.tsx`
: 해당 주(WEEK)의 7개의 일자(DAY)를 출력
- `useWeeksOfMonth.tsx`
: 전달받은 년, 월을 기준으로 42개의 일자(DAY) 리스트를 반환하는 hook
⚠ 매개변수: moment 객체 (년, 월은 필수로 지정)
[EX] `2020년 7월 23일`(moment 객체)를 인자로 전달 받으면,  `2020년 7월 1일`의 **요일 number값을 뺀 일자**부터 42개의 일자 계산
_*** 요일 number : 0(일)~6(토)_

✅ 캘린더 용어
- Year : 년
- Month: 월
- Week: 주
- Day: 일
- WeekDay: 요일
- Date: 일자 ('2020-07-09')

ddb8036631 commented on frontend/src/components/calendar/custom-hook/useWeeksOfMonth.ts:
```diff
@@ -0,0 +1,27 @@
+    const WEEKDAY_1DAY = moment(DATE).day(); // 0(일)-6(월)
+
+    const WEEKS_OF_MONTH = [];
+    let dayPointer = WEEKDAY_1DAY * -1;
+    for (let week = 0; week < 6; week++) { // 6주
+        const WEEK = [];
+        for (let day = 0; day < 7; day++) { // 7일
+            WEEK.push(moment(DATE).add(dayPointer++, 'd'));
+        }
+        WEEKS_OF_MONTH.push(WEEK);
+    }
```
`WEEKDAY_1DAY`라는 변수로 이번 달 1일 이전 일자를 구하는 방법이 참신하네요. 이렇게 포인터로 구현하니까 for문이 훨씬 깔끔해지네요!

ddb8036631 commented on frontend/src/components/calendar/custom-hook/useWeeksOfMonth.ts:
```diff
@@ -0,0 +1,27 @@
+
+export default function useWeeksOfMonth(DATE: moment.Moment) {
+    const [WEEKS_OF_MONTH, setWeeksOfMonth] = useState<moment.Moment[][]>([]);
+    useEffect(() => {
+        setWeeksOfMonth(initWeeksOfMonth(DATE));
+    }, [DATE])
```
`useWeeksOfMonth()`를 호출할 때에만 `DATE`를 매개변수로 받고, `useEffect()`가 `DATE`를 감시하니까, 단 한 번만 실행되는 게 맞나요?
혹시 `MonthContainer.tsx`에서 `WEEKS_OF_MONTH`를 콘솔에 찍었을 때, 무한루프로 출력되는지 확인 가능한가요?

pro-popo replied:
`WEEKS_OF_MONTH`가 무한루프로 출력된다는 건, 
`DATE`가 계속 변하기 때문에 `useWeeksOfMonth` 훅이 무한루프를 돌 것 같다는 얘기가 맞을까요??

일단, 출력을 확인해 보면!
![image](https://user-images.githubusercontent.com/52408122/125209693-9d4e5500-e2d5-11eb-99c7-9c18b8b2e48e.png)

`MonthContainer`(+`useWeeksOfMonth`)는 처음에는 두 번 렌더링이 됩니다.
`useWeeksOfMonth`에서 `WEEKS_OF_MONTH`을 빈 배열(`[]`)로 먼저 초기화를 한 다음에, `useEffect` 훅으로 새로운 값을 저장해서 두 번 렌더링 하는 것 입니다!
그 후, `DATE`가 변할 때 한 번의 렌더링이 발생합니다.

--- 

말씀하신 것처럼, `MonthContainer.tsx`에서 `useWeeksOfMonth` 훅의 `useEffect`는 매개 변수로 전달받은 `DATE`가 변경될 때만 렌더링을 진행할 것 입니다.

`MonthContainer` 컴포넌트는 `YearContainer`컴포넌트에서 2020.1~2020.12월까지 총 12개가 생성됩니다.

✅ `YearContainer.tsx`
```               
{MONTH_LIST.map((month, idx) =>
     <Grid item xs={3} key={idx}>
              <MonthContainer DATE={moment().year(YEAR).month(MONTH)} />
     </Grid>
)}
```

각각의 `MonthContainer` 컴포넌트는 `useWeeksOfMonth`을 호출하는데, 여기서 전부 같은 `useWeeksOfMonth`의 `DATE`를 공유하는 것이 아니라 별도의 메모리에 저장이 됩니다.
`useState`를 여러 번 호출하면 서로 다른 메모리 공간에 저장되는 것처럼, 커스텀 훅도 마찬가지로 서로 다른 메모리에 저장이 됩니다! 
따라서 각각의 `MonthContainer` 컴포넌트의 `DATE`는 서로 영향받지 않아 무한 루프에 빠지지 않는 것 입니다!

설명을 쓰긴 했는데 제가 질문을 제대로 이해한게 맞을까요..?

ddb8036631 replied:
![스크린샷 2021-07-12 오후 9 06 11](https://user-images.githubusercontent.com/50266862/125284912-fd381080-e354-11eb-9c53-b2527676e9bb.png)

네 이 메서드만 가져다 위처럼 제 `Month.tsx` 파일에 렌더링 했을 때에 무한 루프에 빠져서 물어봤습니다.. ㅎㅎ 아직 그 이유를 모르겠네요.

pro-popo replied:
아! 저 왜 무한 루프가 도는지 알 것 같아요!
왜 무한 루프가 돌까, 테스트를 해보니까 `useWeeksOfMonth`의 인자로 `moment()`를 전달해서 문제가 발생하는 것 같습니다!
처음에 `moment()`를 매개변수로 받고, `weeksOfMonth` state가 업데이트되면서 `Month` 컴포넌트를 새로 렌더링 합니다.
여기서 문제는, **컴포넌트를 새로 렌더링하면서 새로운 `moment()`를 전달하기 때문에**, 의존성 배열에 포함한 첫 번째 매개변수, 즉 `moment()`가 변경되었다고 판단해서 또 다시 렌더링을 거치는 것 입니다!

따라서, 인자로 `moment()`를 전달하기 보다는
1. props로 전달
```
const Month = ({today}) => {
  const weeksOfMonth = useWeeksOfMonth(today);
};
```

2. 컴포넌트 외부에 변수 선언
```
const TODAY = moment();
const Month = () => {
  const weeksOfMonth = useWeeksOfMonth(TODAY);
};
```

와 같은 방법을 쓰면 좋을 것 같습니다!

아 참고로, 제가 커밋을 한 번 더 했는데, `useWeeksOfMonth`를 배열 요소로 `moment.Moment` 객체가 아닌 `Moment` 인터페이스를 사용해서 저장했습니다! 무성오빠꺼 코드보고 새로 객체를 만들어서 저장하는게 더 효율적일 것 같아서 바꿔봤어여! 참고해주세여!

ddb8036631 commented on frontend/src/components/calendar/year/Day.tsx:
```diff
@@ -0,0 +1,41 @@
+        <TdDay
+            isCurrentMonth={isCurrentMonth}
+            isToday={isToday}
+            className={isToday ? "today" : "not-today"}
```
`today`와 `not-today` 클래스 명으로 어떤 처리도 하지 않는 것 같습니다. 밑에 `TdDay` 컴포넌트를 봐도,` isCurrentMonth`와 `isToday` 프로퍼티 만으로 스타일링을 하고 있는 것으로 보입니다.
styled-components를 쓰면, 따로 CSS 파일을 만들지 않기에 `className`을 설정하는 게 의미 없다고 생각이 듭니다.

pro-popo replied:
`today`와 `not-today` 클래스명은 `MonthContainer.tsx`의 `TrWeek` 컴포넌트에서 사용합니다!

✅ `MonthContainer.tsx`
```
const TrWeek = styled.tr`
    cursor: pointer;
    & td.not-today:hover {
        background-color: #EEEEEE;
    };
    & td.today:hover {
        background-color: #047769
    };
`;
```
`td`에 직접 hover를 설정하고 싶었는데, `tr`에서 `td`로 접근해야 `hover`를 설정할 수 있어서 선택한 방법입니다.
`Week` 컴포넌트와 `Day` 컴포넌트를 분리해서 보이지 않았나봅니다...!

음, 적당한 방법이 안 떠올라서 선택한 방법이긴 한데 🤔
`className`을 직접 명시하기 때문에, 실제 DOM에서는 `className`이 해쉬로 변환되지는 않습니다.
![image](https://user-images.githubusercontent.com/52408122/125206121-79ccdf80-e2c0-11eb-88da-daf6df23e6c4.png)

다른 방법으로는,
`ref`를 이용해서 자식 컴포넌트를 접근하거나 혹은 직접 `className`을 해쉬로 변환하는 것이 떠오르네요!

혹시 더 좋은 방법이 있을까요?

ddb8036631 replied:
음.. `styled-components`가 해시값으로 컴포넌트 아이디를 사용한다는 걸 처음 알았네요..ㅎㅎ
굳이 클래스명을 해시처리해야 되는 이유가 없다면야 위에 작성하신 방법을 그대로 사용하는 게 좋아 보입니다.

ddb8036631 commented on frontend/src/components/calendar/year/Day.tsx:
```diff
@@ -0,0 +1,41 @@
+    )
+}
+
+type TdDayProps = {
+    isCurrentMonth?: boolean,
+    isToday?: boolean
+}
```
`isCurrentMonth`와 `isToday` 프로퍼티 둘 다 무조건 전달받는 것으로 보이는데, `?:` 를 쓴 이유가 있나요?

pro-popo replied:
`isCurrentMonth`와 `isToday` prop을 선택형(`?:`)으로 작성한 이유는,
두 개의 프로퍼티가 필수 prop이 아니라고 생각했기 때문입니다!

`Day` 컴포넌트를 **재사용**한다고 생각했을 때, 
반드시 `isCurrentMonth`와 `isToday` prop을 추가할 필요가 없다고 판단했습니다!

ddb8036631 commented on frontend/src/components/calendar/year/MonthContainer.tsx:
```diff
@@ -0,0 +1,45 @@
+    DATE: moment.Moment;
+}
+const MonthContainer = ({ DATE }: MonthProps) => {
+    const WEEKS_OF_MONTH = useWeeksOfMonth(DATE);
```
혹시 `WEEKS_OF_MONTH`를 출력했을 때, 콘솔에서 무한정으로 출력되는 지 확인 가능할까요..?

pro-popo replied:
위에 설명 추가했습니다! 😆

ddb8036631 commented:
커멘트 달아준 ✅ 캘린더 용어가 `moment`모듈에서 사용되는 메서드 설명이 맞나요? 그럼 day()랑 weekday() 둘 다 요일을 의미하는 거 같던데.. 차이점이 더 있나요?
또, 코드 보면서 `moment().year(2021).month(0)`식으로 파라미터 주면 해당되는 **년**, **달**로 설정되는 것 알게 되었습니다.

pro-popo commented:
아 제가 적어놓은 캘린더 용어는, 실제 용어를 영어로 표현한 단어고, `moment` 모듈 기준이 아닙니다...😅
엑셀 썼을 때 생각나서 아래처럼 정한건데
```
date = 일자
day = 일
week of day = 요일
```
생각해보니까, `moment`에서는 date = 일, (day | weekday) = 요일. 이렇게 사용하고 있네요...
 `moment` 기준으로 수정하겠습니다!

```
date = 일
day = 요일
```

pro-popo committed:
refactor: 컴포넌트명, 변수명 수정 및 type Moment 설정
- 컴포넌트명, 변수명 수정
- Week 캘린더 컴포넌트화
- weeksOfMonth 원소 타입 수정 => moment.Moment 객체 대신 interface Moment로 설정

pro-popo commented:
✅ 컴포넌트명 및 변수명 수정
- 컴포넌트명이랑 변수명이 일관성 없는 것 같아서 전반적으로 수정했습니다!
1. 컴포넌트 외부에 있는 변수명(상수)만 대문자로 표기
2. 재사용 불가능한 컴포넌트를 `Container`라고 명명하는게 나은 것 같아서 전부 수정

✅ `useWeeksOfMonth` 수정
- `useEffect` 대신 `useLayoutEffect` 사용 
=> 새로고침 시, 타이틀과 날짜들이 끊겨서 나와서 수정했습니다!

- `moment.Moment` 객체 대신, 새로운 `Moment interface`로 생성해서 저장
일 년 단위로 날짜를 가져오면 500개 이상의 `moment.Moment` 객체를 생성하는데, 사용하는 데이터의 양에 비해 크기가 너무 크다고 판단했습니다!
![image](https://user-images.githubusercontent.com/52408122/125214490-6fc3d480-e2f2-11eb-9c5f-f7b6027591df.png)
그래서, 필요한 정보만 모아서 `Moment interface`로 저장했습니다!
```
export interface Moment {
    year: number;
    month: number;
    date: number;
}
```

pro-popo committed:
refactor: CurrentMonthContext로 Month 캘린더의 month value 관리 및 전달

pro-popo commented:
✅ `CurrentMonthContext` 활용
`MonthCalendar`의 현재 `month value`를 자식 컴포넌트에게 전달하기 위해, `useWeeksOfMonth` Hook에서 `CurrentMonthContext`를 생성했습니다!
